### PR TITLE
IGNITE-21171 SQL Calcite: Fix column nullability for data types with …

### PR DESCRIPTION
**Pull request by alex-plekhanov**
…precession or scale

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java [hunk 1]
```diff
@@ -496,6 +496,35 @@ public void createTableAsSelectFromDistributedTable() {
...
+                sql("CREATE TABLE my_table (id INT PRIMARY KEY, val " + dataType + ')');
+                sql("INSERT INTO my_table(id, val) VALUES (0, NULL)");
+            }
+            finally {
+                sql("DROP TABLE my_table");
+            }
+
+            try {
+                sql("CREATE TABLE my_table (id INT PRIMARY KEY, val " + dataType + " NOT NULL)");
+                assertThrows("INSERT INTO my_table(id, val) VALUES (0, NULL)", IgniteSQLException.class,
```
Comment (Vladsz83) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java:519]: Does this duplicate alterTableAddNotNullColumn() or alterTableDropAddColumn() ?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java [hunk 2]
```diff
@@ -180,14 +180,22 @@ private static RelDataType createRowType(IgniteTypeFactory typeFactory, List<Rel
...
         if (javaType.getSqlTypeName().allowsPrecScale(true, true) &&
-            (precision != RelDataType.PRECISION_NOT_SPECIFIED || scale != RelDataType.SCALE_NOT_SPECIFIED))
-            return typeFactory.createSqlType(javaType.getSqlTypeName(), precision, scale);
+            (precision != RelDataType.PRECISION_NOT_SPECIFIED || scale != RelDataType.SCALE_NOT_SPECIFIED)) {
+            return typeFactory.createTypeWithNullability(
+                typeFactory.createSqlType(javaType.getSqlTypeName(), precision, scale), nullability);
+        }
 
-        return sqlType(typeFactory, javaType);
+        return typeFactory.createTypeWithNullability(sqlType(typeFactory, javaType), nullability);
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java:198]: Do we have any test for this?

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/QueryMetadataIntegrationTest.java [hunk 3]
```diff
@@ -42,13 +42,14 @@ public class QueryMetadataIntegrationTest extends AbstractBasicIntegrationTest {
     @Test
     public void testJoin() throws Exception {
         executeSql("CREATE TABLE tbl1 (id DECIMAL(10, 2), val VARCHAR, val2 BIGINT, ts TIMESTAMP(14), PRIMARY KEY(id, val))");
-        executeSql("CREATE TABLE tbl2 (id DECIMAL(10, 2), val VARCHAR, val2 BIGINT, ts TIMESTAMP(14), PRIMARY KEY(id, val))");
+        executeSql("CREATE TABLE tbl2 (id DECIMAL(10, 2) NOT NULL, val VARCHAR, val2 BIGINT, ts TIMESTAMP(14), " +
```
Comment (Vladsz83) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/QueryMetadataIntegrationTest.java:45]: Suggestion: + a similar test to cover H2/GridQueryProcessor queries like `SqlFieldsQueryUsageTest`. WDYT?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java (hunk 2 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java:198]: `alterTableAddNotNullColumn` for example. There was another error message in this test, because null values were allowed for type and value was checked later for nullability (during insert). Now type doesn't allow null values and error message changed in test.

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/QueryMetadataIntegrationTest.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/QueryMetadataIntegrationTest.java:45]: 1. `SqlFieldsQueryUsageTest` was created to test DDL when DDL was not supported by calcite, now this test make no sense and can be deleted.
2. I beleive that H2-engine tests for nullability already exists (looks like `IgniteSqlNotNullConstraintTest` checks nullability)

**Comment by Vladsz83** [comment]
LGTM

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java (hunk 1 above)
Comment (alex-plekhanov) [reply modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/TableDdlIntegrationTest.java:519]: There are different commands: `AlterTableAddCommand` and `CreateTableCommand` and nullability flag is processed by them differently (see `DdlCommandHandler#handle`). I think it's better to check that nullablility is not lost in both cases.
